clio-linkchecking/src/main/java/eu/europeana/clio/linkchecking/LinkCheckingEngine.java

```diff
@@ -95,38 +131,62 @@ public void createRunWithUncheckedLinksForDataset(String datasetId) throws ClioE
    * @throws ClioException In case of a problem with accessing or storing the required data.
    */
   public void performLinkCheckingOnAllUncheckedLinks() throws ClioException {
-    final LinkDao linkDao = new LinkDao(properties.createPersistenceConnection());
-    try (final LinkChecker linkChecker = createLinkChecker()) {
-      boolean linksChecked;
-      do {
-        linksChecked = checkNextLinkIfAvailable(linkChecker, linkDao);
-      } while (linksChecked);
+    try (
+        final ClioPersistenceConnection databaseConnection =
+            properties.getPersistenceConnectionProvider().createPersistenceConnection();
+        final LinkChecker linkChecker = createLinkChecker()) {
+      final LinkDao linkDao = new LinkDao(databaseConnection);
+      try (final StreamResult<Link> linksToCheck = linkDao.getAllUncheckedLinks()) {
+        ParallelTaskExecutor.executeAndWait(linksToCheck.get(),
+                link -> performLinkCheckingOnUncheckedLink(linkChecker, linkDao, link),
+                properties.getLinkCheckingRunExecuteThreads());
```

Comment:
In fact there is another issue: the code is not properly synchronized. I will look at this more carefully and make some changes.

Reply:
I have made the code thread-safe. I think that for the issue you originally raised there is no easy solution given the current implementation. It could be its own ticket. But for now I put a comment in the config file: if you just choose sufficiently many execution threads compared to the number of records we poll per dataset we should not have a really big problem.